File: WeatherFrontend/Services/WeatherServiceImpl.cs
using Google.Protobuf;

namespace WeatherFrontend.Services;

public class WeatherServiceImpl
{
    private readonly WeatherFrontend.WeatherService.WeatherServiceClient _client;

    public WeatherServiceImpl(WeatherService.WeatherServiceClient client)
    {
        _client = client;
    }

    public async Task<ActualTemperatureResponse> GetActualTemperature(string town)
    {
        return await _client.GetActualTemperatureAsync(new ActualTemperatureRequest()
        {
            Location = new Location()
            {
                Town = town
            }
        });
    }

    public async Task<ForecastResponse> GetForecast(float logtitude, float latitude, string startDate, string endDate, int forecastDays)
    {
        return await _client.GetForecastAsync(new ForecastRequest()
        {
            Longitude = logtitude,
            Latitude = latitude,
            Timezone = "GMT",
            StartDate = startDate,
            EndDate = endDate,
            ForecastDays = forecastDays
        });
    }

    public async Task<RecognizeWeatherResponse> RecognizeImage(Stream stream)
    {
        return await _client.RecognizeWeatherAsync(new RecognizeWeatherRequest()
        {
            Image = await ByteString.FromStreamAsync(stream)
        });
    }
}

File: DesktopKotlinWeather/src/main/kotlin/service/WeatherServiceImpl.kt
package service

import com.google.protobuf.ByteString
import weather.Weather
import weather.WeatherServiceGrpcKt


class WeatherServiceImpl(private val stub: WeatherServiceGrpcKt.WeatherServiceCoroutineStub) {

    suspend fun getActualTemperature(town: String): Weather.ActualTemperatureResponse {
        val request = Weather.ActualTemperatureRequest.newBuilder()
            .setLocation(Weather.Location.newBuilder().setTown(town))
            .build()
        return stub.getActualTemperature(request)
    }

    suspend fun getForecast(
        latitude: Float,
        longitude: Float,
        startDate: String,
        endDate: String,
        forecastDays: Int,
        forecastHours: Int
    ): Weather.ForecastResponse {
        val request = Weather.ForecastRequest.newBuilder()
            .setLatitude(latitude)
            .setLongitude(longitude)
            .setTimezone("GMT")
            .setStartDate(startDate)
            .setEndDate(endDate)
            .setForecastDays(forecastDays)
            .setForecastHours(forecastHours)
            .build()
        return stub.getForecast(request)
    }

    suspend fun recognizeImage(imageBytes: ByteArray): Weather.RecognizeWeatherResponse {
        val request = Weather.RecognizeWeatherRequest.newBuilder()
            .setImage(ByteString.copyFrom(imageBytes))
            .build()
        return stub.recognizeWeather(request)
    }
}